File: tests/julia/test_islm.jl
using Test
using MacroAIComp

@testset "ISLM Tests" begin
    @testset "Steady State Default Calibration" begin
        params = default_calibration(ISLMParams)
        ss = steady_state(params)

        @test ss["Y"] ≈ 2000.0 atol=1e-6
        @test ss["P"] ≈ 81.0 atol=1e-6
        @test ss["i"] ≈ 2.0 atol=1e-6
        @test ss["Yd"] ≈ 2000.0 atol=1e-6
        @test ss["dP"] ≈ 0.0 atol=1e-6
        @test ss["dY"] ≈ 0.0 atol=1e-6
    end

    @testset "System Dynamics at Steady State" begin
        params = default_calibration(ISLMParams)
        ss = steady_state(params)
        state = [ss["Y"], ss["P"]]

        derivatives = system_dynamics(state, params, 0.0)

        @test derivatives[1] ≈ 0.0 atol=1e-6  # dY/dt
        @test derivatives[2] ≈ 0.0 atol=1e-6  # dP/dt
    end

    @testset "Monetary Shock" begin
        params = default_calibration(ISLMParams)
        # Shock: money supply increases from 100 to 101.
        # Since Julia structs are immutable, we reconstruct it.
        params_shocked = ISLMParams(
            params.theta,
            params.psi,
            params.beta1,
            params.mi,
            params.ni,
            params.beta0,
            101.0,
            params.ypot0
        )
        ss_new = steady_state(params_shocked)

        @test ss_new["Y"] ≈ 2000.0 atol=1e-6
        @test ss_new["i"] ≈ 2.0 atol=1e-6
        @test ss_new["P"] ≈ 82.0 atol=1e-6
    end
end

File: tests/python/test_islm.py
import pytest
import numpy as np
from macroaicomp.models.islm import default_calibration, steady_state, system_dynamics


def test_steady_state_default_calibration():
    """Test that the steady state matches the Oracle values for default calibration."""
    params = default_calibration()
    ss = steady_state(params)

    # Values from Appendix D (MATLAB) and Oracle
    assert pytest.approx(ss["Y"], 1e-6) == 2000.0
    assert pytest.approx(ss["P"], 1e-6) == 81.0
    assert pytest.approx(ss["i"], 1e-6) == 2.0
    assert pytest.approx(ss["Yd"], 1e-6) == 2000.0
    assert pytest.approx(ss["dP"], 1e-6) == 0.0
    assert pytest.approx(ss["dY"], 1e-6) == 0.0


def test_system_dynamics_at_steady_state():
    """Derivatives should be zero at the steady state."""
    params = default_calibration()
    ss = steady_state(params)
    state = np.array([ss["Y"], ss["P"]])

    derivatives = system_dynamics(0.0, state, params)

    assert pytest.approx(derivatives[0], 1e-6) == 0.0  # dY/dt
    assert pytest.approx(derivatives[1], 1e-6) == 0.0  # dP/dt


def test_shock_monetary():
    """Test that after a monetary shock, the new steady state is correct."""
    params = default_calibration()
    # Shock: money supply increases from 100 to 101
    params.m0 = 101.0

    ss_new = steady_state(params)

    # Output and interest rate shouldn't change in the long run (money neutrality)
    assert pytest.approx(ss_new["Y"], 1e-6) == 2000.0
    assert pytest.approx(ss_new["i"], 1e-6) == 2.0

    # Prices should increase by the same amount as money supply (101 - 100 = 1) -> 81 + 1 = 82
    assert pytest.approx(ss_new["P"], 1e-6) == 82.0
